dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/AbstractTrackerValidationHook.java

```diff
@@ -0,0 +1,393 @@
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1010 )
+                    .withObject( te ) );
+                return;
+            }
+
+            OrganisationUnit organisationUnit = PreheatHelper.getOrganisationUnit( bundle, te.getOrgUnit() );
+
+            if ( organisationUnit == null )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1011 )
+                    .withObject( te )
+                    .addArg( te.getOrgUnit() ) );
+            }
+
+        }
+        else if ( bundle.getImportStrategy().isUpdate() )
+        {
+            TrackedEntityInstance tei = PreheatHelper.getTrackedEntityInstance( bundle, te.getTrackedEntity() );
+
+            if ( tei.getOrganisationUnit() == null )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1011 )
+                    .withObject( te )
+                    .addArg( tei ) );
+            }
+        }
+    }
+
+    protected boolean textPatternValueIsValid( TrackedEntityAttribute attribute, String value, String oldValue )
+    {
+        return Objects.equals( value, oldValue ) ||
+            TextPatternValidationUtils.validateTextPatternValue( attribute.getTextPattern(), value ) ||
+            reservedValueService.isReserved( attribute.getTextPattern(), value );
+    }
+
+    private boolean validateTextPattern( ValidationErrorReporter errorReporter, Attribute attr,
+        TrackedEntityAttribute teAttr,
+        TrackedEntityAttributeValue teiAttributeValue )
+    {
+        if ( teAttr.isGenerated() && teAttr.getTextPattern() != null )
+        //&& ??? !importOptions.isSkipPatternValidation()
+        // MortenO: How should we deal with this in the new importer?
+        {
+            String oldValue = teiAttributeValue != null ? teiAttributeValue.getValue() : null;
+            if ( !textPatternValueIsValid( teAttr, attr.getValue(), oldValue ) )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1008 )
+                    .addArg( attr.getValue() ) );
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private boolean validateFileNotAlreadyAssigned( ValidationErrorReporter errorReporter, Attribute attr,
+        TrackedEntityInstance tei )
+    {
+        boolean attrIsFile = attr.getValueType() != null && attr.getValueType().isFile();
+
+        if ( attrIsFile && tei != null )
+        {
+            List<String> existingValues = new ArrayList<>();
+
+            tei.getTrackedEntityAttributeValues().stream()
+                .filter( attrVal -> attrVal.getAttribute().getValueType().isFile() )
+                .forEach( attrVal -> existingValues.add( attrVal.getValue() ) );
+
+            FileResource fileResource = fileResourceService.getFileResource( attr.getValue() );
+            if ( fileResource != null && fileResource.isAssigned() && !existingValues.contains( attr.getValue() ) )
```

Comment:
This might result in the wrong result.

A file-resource represents a single file, however, a file-resource also represents a single value. That means a file-resource can only ever be connected to a single value.

With this code, we are saying that a file-resource can belong to multiple values, as long as they are owned by the same tei. The only change I see needed to keep the logic is to make existingValues solely contain the value of the specific attribute we are looking at.

Reply:
I have tried to implement this now, please check to confirm.

Reply:
@netroms I dont see a change in the code here? I think it should be fixed if you add another filter to line 275, which only returns the attribute with the attr.uid. that way we should only get 1 attributeValue, which should be the one we would allow if we are updating.

Reply:
@stian-sandvold Sorry, I think the change was not in the last commti. Should be there now.
Looks like this: line 272: filter( attrVal -> attrVal.getAttribute().getUid()

Reply:
I still dont see it 🤷‍♂ but it sounds like you got the idea, so just make sure its there for the final PR :)

Reply:
https://github.com/dhis2/dhis2-core/blob/98257401e5c8cbfe0692b733a142db5255005da2/dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/AbstractTrackerValidationHook.java